src/Common/PSR16Util.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+namespace Vectorface\Cache\Common;
+
+use DateInterval;
+use DateTime;
+use Traversable;
+use Vectorface\Cache\Exception\CacheException;
+use Vectorface\Cache\Exception\InvalidArgumentException as CacheArgumentException;
+
+
+/**
+ *
+ */
+trait PSR16Util
+{
+    /**
+     * The DateTime implementation to use
+     *
+     * @var string
+     */
+    private static $dateTimeClass = DateTime::class;
+
+    /**
+     * Enforce a fairly standard key format
+     *
+     * @param $key
+     * @return mixed Returns the key, if valid
+     * @throws CacheArgumentException Thrown if the key is not a legal value
+     */
+    protected function key($key)
+    {
+        if (is_scalar($key)) {
+            return $key;
+        }
+
+        throw new CacheArgumentException("key is not a legal value");
+    }
+
+    /**
+     * Enforce fairly standard key formats on an iterable of values
+     * @param iterable $values
+     * @return iterable The values array
+     * @throws CacheArgumentException Thrown if any of the keys is not a legal value
+     */
+    protected function values($values) {
+        foreach ($values as $key => $value) {
+            $key = $this->key($key); // Checks the key
+        }
+        return $values;
+    }
+
+    /**
+     * Enforce a fairly standard key format on an array or Traversable of keys
+     *
+     * @param iterable $keys
+     * @return mixed Returns the key, if valid
+     * @throws CacheArgumentException Thrown if any of the keys is not a legal value
+     */
+    protected function keys($keys)
+    {
+        if (!is_array($keys) && !($keys instanceof Traversable)) {
+            throw new CacheArgumentException("keys must be provided as an array or a Traversable");
+        }
+
+        foreach ($keys as &$key) {
+            $key = $this->key($key);
+        }
```

Comment:
I'm not sure if there's any value in using a `&` and assigning here. Can just `$this->key($key)` and ignore the return value.

I figure your intent is to allow `key()` to be overriden by an implementing class though.

Also I'm not sure if there's any guarantee that Traversable elements are mutable. I'm thinking of the situation where a `Collection` class of keys is passed in or something.

Reply:
Good point about traversables. I'll have to add some testing for that too because this is likely not fully compliant with PSR-16 if it doesn't handle Traversables correctly.

I had initially included a key validation and modification callback mechanism that I decided not to include because it would be feature creep. I think all the classes use the key/keys/value/values methods in such a way that it should work once that's in place, *except* that some end up calling key/keys twice so modification won't work without further fixes.